### ISSUE TITLE
`qjit(static_argnums=...)` fails when the marks static argument has a default value

### ISSUE THREAD

**Issue by paul0403** [PennyLaneAI/catalyst#1163]
```python3
@qjit(static_argnums=[1])
def f(y, x=9):
    if x < 10:
        return x + y
    return 42000


res = f(20)
print(res)
```

```
Traceback (most recent call last):
  File "/home/paul.wang/catalyst_new/catalyst/static_argname.py", line 24, in <module>
    res = f(20)
  File "/home/paul.wang/.local/lib/python3.10/site-packages/pennylane/logging/decorators.py", line 61, in wrapper_entry
    return func(*args, **kwargs)
  File "/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/jit.py", line 513, in __call__
    requires_promotion = self.jit_compile(args, **kwargs)
  File "/home/paul.wang/.local/lib/python3.10/site-packages/pennylane/logging/decorators.py", line 61, in wrapper_entry
    return func(*args, **kwargs)
  File "/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/jit.py", line 584, in jit_compile
    self.jaxpr, self.out_type, self.out_treedef, self.c_sig = self.capture(
  File "/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/debug/instruments.py", line 143, in wrapper
    return fn(*args, **kwargs)
  File "/home/paul.wang/.local/lib/python3.10/site-packages/pennylane/logging/decorators.py", line 61, in wrapper_entry
    return func(*args, **kwargs)
  File "/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/jit.py", line 631, in capture
    verify_static_argnums(args, self.compile_options.static_argnums)
  File "/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/tracing/type_signatures.py", line 113, in verify_static_argnums
    raise CompileError(msg)
catalyst.utils.exceptions.CompileError: argnum 1 is beyond the valid range of [0, 1).

```

Note that jax works:
```python
@partial(jax.jit, static_argnums=[1])
def f(y, x=9):
    if x < 10:
        return x + y
    return 42000


res = f(20)
print(res)
```
29
```

```

**Comment by paul0403** [comment]
I think this is because the `static_argnum` tracks the argument indices at the call, not at the definition

**Comment by AniketDalvi** [comment]
Hi! I am Aniket, a PhD candidate at Duke interviewing for a role on the compiler team. I was given this issue as a technical challenge. Apart from the details mentioned in this issue, are there any other pointers that might help me tackle this issue? Thank you!

**Comment by paul0403** [comment]
Hi @AniketDalvi , do you have a more specific question in mind? I would love to give out more pointers and resolve any confusion you may have!

**Comment by AniketDalvi** [comment]
Hi! So for installation - it says to download the PyPI wheel. Which version should I be downloading the wheel for? `pip` says there are 3 available versions - `0.1.0, 0.1.1 and 0.1.2`.

**Comment by erick-xanadu** [comment]
Hey @AniketDalvi, I am not too sure why your pip says that there are only 3 versions available.[ We have version 0.8.1 in pypi](https://pypi.org/project/PennyLane-Catalyst/) and will be releasing version 0.9.0 soon.

**Comment by AniketDalvi** [comment]
Okay I am just going to download the `.whl` file from the above link. I will then extract it in the `frontend` directory as was directed in the installation instructions

**Comment by erick-xanadu** [comment]
> Hi! Okay, so I have followed the instruction and seemed to have successfully installed the repository. Is there a quick sanity check experiment/file I can run to verify the installation?

You can run the tests. `pytest frontend/test/pytest -n auto`. I think it is possible that some debugging tests fail in your machine as they expect a specific path. But if the vast majority are passing, then it should be good.

**Comment by AniketDalvi** [comment]
Running pytest gave me an error stating that there was an interpreter mismatch. The interpreter is Python 3.10 while the package is compatible only with 3.12

**Comment by erick-xanadu** [comment]
> Running pytest gave me an error stating that there was an interpreter mismatch. The interpreter is Python 3.10 while the package is compatible only with 3.12

When you download via pip, pip enforces this compatibility check, but manually downloading it you need to make sure to download the appropriate one for you. See here for a list of several wheels with different python version compatibility.

Maybe it would be easier to install from source? It just takes a long time to build LLVM initially.

**Comment by AniketDalvi** [comment]
Understood, that makes sense. I might re-try it with a different wheel with a compatible python version. If not, I will resort to installing from source.

**Comment by paul0403** [comment]
> Okay when trying to run the `make frontend` command with the new wheel, I get this error - `error: command '/usr/bin/g++' failed with exit code 1`. I have tried all solutions that stack overflow has to offer, but to no avail. Any thoughts on what this could be?

There's some required packages before building Catalyst. Maybe some of them are missing? See the [build from source guide](https://docs.pennylane.ai/projects/catalyst/en/stable/dev/installation.html#minimal-building-from-source-guide)

**Comment by AniketDalvi** [comment]
Yup installed all the required packages, but the error persists. I am now just going to build from source instead.

**Comment by paul0403** [comment]
> Yup installed all the required packages, but the error persists. I am now just going to build from source instead.

To avoid all package version issues, I also recommend using a fresh virtual environment when developing, e.g.
```
python3 -m venv pyenv
source ./pyenv/bin/activate
```

after which you can pip install all the requirements and `make all` from source.

**Comment by AniketDalvi** [comment]
Okay I seemed to have gotten it to work from source. Most tests pass, some are skipped, and 4 debugging tests fails (as @erick-xanadu said is expected). I am running all of this from with within a conda environment on a linux machine.

**Comment by erick-xanadu** [comment]
> Okay I seemed to have gotten it to work from source. Most tests pass, some are skipped, and 4 debugging tests fails (as @erick-xanadu said is expected). 

Yeah, a bunch are skipped, 4 failing ones.

>  I am running all of this from with within a conda environment on a linux machine.

Awesome! We don't normally use conda so I am happy to hear this worked for you :)

**Comment by AniketDalvi** [comment]
Hi! So from my initial analysis, I traced the issue down to this check that throws an exception - https://github.com/PennyLaneAI/catalyst/blob/75dc517a6e2f2583da890b9e8198241933a7aef5/frontend/catalyst/tracing/type_signatures.py#L121.

It appears that it checks the index used to specify the static argument with the number of `args` passed to the function. However, in the case of a default parameter, the static argument index is likely to be greater than `len(args)` as the default parameter may not be passed when the function is called. Considering this, my initial proposal for a solution would be to compare the static argument index to the number of arguments in the function signature, as opposed to the number in the function itself.

Would like to understand your thoughts on this!

**Comment by paul0403** [comment]
> Hi! So from my initial analysis, I traced the issue down to this check that throws an exception -
> 
> https://github.com/PennyLaneAI/catalyst/blob/75dc517a6e2f2583da890b9e8198241933a7aef5/frontend/catalyst/tracing/type_signatures.py#L120
> 
> .
> It appears that it checks the index used to specify the static argument with the number of `args` passed to the function. However, in the case of a default parameter, the static argument index is likely to be greater than `len(args)` as the default parameter may not be passed when the function is called. Considering this, my initial proposal for a solution would be to compare the static argument index to the number of arguments in the function signature, as opposed to the number in the function itself.
> 
> Would like to get your thoughts on this!

Hi! Usually what we do for these challenges is you can fork the catalyst repo, push your changes, and open a PR. It will be easier to review. 

It's good if it turns out to be a simple fix, but one thing I'm afraid of is whether loosening the verification would allow in some errors. Can you test the frontend test suite to make sure this does not happen? You can run `make test-frontend` in the root catalyst directory.

**Comment by AniketDalvi** [comment]
Hi! Okay that sounds good. I am working off of a my branch. Does that work, or does it have to be a fork?

**Comment by erick-xanadu** [comment]
It doesn't really matter how you develop, as long as you are able to push a pull request :). I think to do that, you do need a fork. But you can always just add a new remote to your local git workspace.

```bash
git clone $pennylane/catalyst
# work on the issue
# fork $pennylane/catalyst to $yourrepo/catalyst
git remote add myrepo $yourrepo/catalyst
git push myrepo $yourbranch
# open a PR from $yourrepo/catalyst to $pennylane/catalyst
```
